### ISSUE TITLE
Exceptions in ChatQnA services logs when it gets larger number of requests

### ISSUE THREAD

**Issue by eero-t** [opea-project/GenAIExamples#469]
**Setup**

These errors happen with v0.7 ChatQnA Xeon installation [1], but e.g. updating to TEI services from `1.2-cpu` version to latest `1.5-cpu`, and and TGI service from `1.4` version to latest `2.2` did help.

[1] https://github.com/opea-project/GenAIExamples/tree/v0.7/ChatQnA/kubernetes/manifests

**Use-case**

Constantly stress ChatQnA  `chaqna-xeon-backend-server-svc` service endpoint by sending it large[2] number of queries in parallel.

[2] compared to actual capacity of the service. E.g. 8 queries in parallel for service running on IceLake Xeon.

**Actual outcome**

Occasionally:
* Process sending it queries gets `unexpected EOF` error (i.e. service reply ended before specified `Content-Length`)
* Exception in logs from few of the service pods

For the exeption details, see the attachments:
* [`chaqna-xeon-backend-server-svc`](https://github.com/user-attachments/files/16393365/chatqna-exceptions.txt)
* [`reranking-svc`](https://github.com/user-attachments/files/16393416/reranking-exceptions.txt)
* [`llm-svc`](https://github.com/user-attachments/files/16393368/llm-exceptions.txt)

Logs for pods of `embedding-svc`, `redis-vector-db`, `retriever-svc`, `tei-embedding-svc`, `tei-reranking-svc` and `tgi-svc` services did not show any exceptions or other errors.

`unexpected EOF` error can happen before  `chaqna-xeon-backend-server-svc` replies with the first token, or after it has already provided 100-200 tokens for the reply.

**Expected outcome**

* Services handle common exceptions gracefully; shortly log the error and tell the caller that now is not a good time to do queries (return e.g. `503 Service Unavailable`), instead of spamming log with exception and "crashing" the reply connection

* Have some kind of rate-limiting for `chaqna-xeon-backend-server-svc`, so that if it gets too many replies before earlier ones have been processed promptly enough, it will start replying `503` pre-emptively[2]  (instead of making the situation worse by trying to process all requests although it has no capacity for it currently, and then failing in middle)

Note: Rate-limiting helps when service scale-up is slow (TGI pod may take minutes from startup until it's ready to respond), and once service has been scaled as up as it can go.  But it needs to be done so that it does not prevent scale-up, or cause too much fluctuation for it.

(No comment on whether that should be implemented in  `chaqna-xeon-backend-server-svc` itself, or in some load-balancer in front of it.)

**Comment by eero-t** [comment]
FYI: I have (k8s) readiness probes on TGI and TEI services, because otherwise things fail due to (k8s) service endpoint sending traffic to recently scaled up TGI instances, although TGI is not yet ready to accept requests.

However, sometimes TGI goes to non-ready state some time later (e.g. when there are multiple instances on same Xeon node, and all of them are stressed at the same time).  There are no warnings in logs, no crashes. TGI just stops responding, k8s marks them as non-ready, and does not route traffic to them (for a while).   This is likely cause at least for some of those exceptions in the services using TGI.

**Comment by YuningQiu** [comment]
Hello, thanks for bringing up this issue. The latest OPEA v0.8 has been released, and the latest version of ChatQnA has TEI services as [1.5-cpu](https://github.com/opea-project/GenAIExamples/blob/a2437e83e74247d605b67ce4a28bb3ddfed1e86f/ChatQnA/docker/xeon/compose.yaml#L30)  and TGI service as [v2.1.0](https://github.com/opea-project/GenAIExamples/blob/a2437e83e74247d605b67ce4a28bb3ddfed1e86f/ChatQnA/docker/xeon/compose.yaml#L115). 

Could you please give a try on v0.8 ChatQnA and let us know if this issue is still bothering you? Thanks a lot!

**Comment by xiguiw** [comment]
@eero-t 
Current test 32 requests with no such issues.

Is this issue still valid at your side?
Could we close it?

**Comment by eero-t** [comment]
> Current test 32 requests with no such issues.

I don't think that is stressing it enough, if you do not see even Timeout issues.  I'm seeing these kind of issues every time I stress any of these services, either on CPU or Gaudi.  Either test does not use enough parallel requests or token counts are too low (mixing in requests with invalid content may also help seeing exceptions).

Btw. What `megaservice_request_pending` values ChatQnA is reporting for your test-runs, from its `/metrics` endpoint?  Is it about 32 _or smaller_?

> Is this issue still valid at your side?

Yes, but it's not Xeon, aipc or even ChatQnA specific.

E.g. this ticket shows exceptions seen in DocSum app log: https://github.com/opea-project/GenAIExamples/issues/1481

To re-iterate the point from the original bug description... The expectation is that well written code catches exceptions and provides user a readable description of the caught issue, instead of connection just breaking when backend service fails to return an answer.

**Comment by xiguiw** [comment]
@eero-t 
I have only on machine, so I could not send large number of requests.

Did you deploy on K8S with multiple machines?  What's the rough number and the platform (how many machines in the system) of the request you tried?

**Comment by eero-t** [comment]
> [@eero-t](https://github.com/eero-t) I have only on machine, so I could not send large number of requests.
> 
> Did you deploy on K8S with multiple machines? What's the rough number and the platform (how many machines in the system) of the request you tried?

@xiguiw Sorry for the late reply.  Just tried with latest ChatQnA, and single node is enough.

I used changing number (64-256) of parallel (non-identical) requests being constantly sent.  vLLM replica count was scaling up and down between 1-2 instances with HPA, but manual scaling should work too.

ChatQnA log with "OK" lines removed is attached, it shows multiple exceptions: 
[chatqna-exceptions.txt](https://github.com/user-attachments/files/19452515/chatqna-exceptions.txt)

**Comment by xiguiw** [comment]
> > [@eero-t](https://github.com/eero-t) I have only on machine, so I could not send large number of requests.
> > Did you deploy on K8S with multiple machines? What's the rough number and the platform (how many machines in the system) of the request you tried?
> 
> [@xiguiw](https://github.com/xiguiw) Sorry for the late reply. Just tried with latest ChatQnA, and single node is enough.
> 
> I used changing number (64-256) of parallel (non-identical) requests being constantly sent. vLLM replica count was scaling up and down between 1-2 instances with HPA, but manual scaling should work too.
> 
> ChatQnA log with "OK" lines removed is attached, it shows multiple exceptions: [chatqna-exceptions.txt](https://github.com/user-attachments/files/19452515/chatqna-exceptions.txt)

@eero-t 

Thanks for your detail information.
So you send parallel (non-identical) requests, the number of requests varies from 64 to 256.
What's the frequency you send out the request? 


I benchmarked vLLM on Xeon GNR, with BKC in August last year.
The batch is < 90 (vLLM only, no retriver/raranking etc.). 

I don't think ChatQnA can process 90 request on single node Xeon.
It's a question how many requests can be processed by Xeon CPU. There should be a limited number.

The problem is what's the expected behavior shoud be if the requests beyond the hardware capabilities.

What's you expectation about this issue?

**Comment by eero-t** [comment]
> So you send parallel (non-identical) requests, the number of requests varies from 64 to 256.
> What's the frequency you send out the request?

64-256 parallel requests means that there's **constantly** that number of unfinished requests.  When one request completes, next one is sent immediately.

> I don't think ChatQnA can process 90 request on single node Xeon.

Sorry, I meant single 8xGaudi2 node.

When I filed this last summer, I was still using TGI on CPU, but I switched to Gaudi2 already last fall, and this year the LLM backend is vLLM.  The ChatQnA megaservice exceptions have happened with _all_ of those variants.

=> it does not really depend on what the backend is or where it runs, as long as ChatQnA is stressed enough that LLM instance(s) start rejecting requests, or are scaled up and down.

**Comment by xiguiw** [comment]
@eero-t 

Thank you for your feedback.

> > So you send parallel (non-identical) requests, the number of requests varies from 64 to 256.
> > What's the frequency you send out the request?
> 
> 64-256 parallel requests means that there's **constantly** that number of unfinished requests. When one request completes, next one is sent immediately.

I wonder if I can reproduce at my side and look into it.
Would you let me know how you did this test? Can you share the BKM to set up the environment?

Something I want to check:
1. How many cards are used to run vLLM-server?
   LLM can use multiple cards, though it does not likely the one blocks the pipeline.

2. TEI-Embedding is running on CPU
https://github.com/opea-project/GenAIExamples/blob/063547fb66ef20badc1e9852b89069867d89ceff/ChatQnA/docker_compose/intel/hpu/gaudi/compose.yaml#L36

It's possible blocked by CPU.
 a) Need to me check if TEI requests can be processed in batch.
 b) What's the utilization of CPUs? (100% buys?)

3.  TEI reranking running on Gaudi.
 But only one 1 card is allocated.
   Is it possible blocked by Reranking?

https://github.com/opea-project/GenAIExamples/tree/main/ChatQnA/docker_compose/intel/hpu/gaudi#composeyaml---default-deployment

**Comment by eero-t** [comment]
This issue is also visible in CI test failures, like ones currently with DocSum: https://github.com/opea-project/GenAIExamples/issues/1936

**Comment by yinghu5** [comment]
Hi @eero-t 
The issue #1936  was resolved after upgrade the langchain[Remind], please feel free to try the latest version and see if the issue can be resolved.   thank you!

**Comment by eero-t** [comment]
Gaudi vLLM has a bug that when it's scaled down, it will just stall and eventually die (= break request connnections), instead of finishing the queries it's already processing + terminating gracefully. See https://github.com/HabanaAI/vllm-fork/issues/1426

=> Easy way to currently see the exceptions in application log, is scaling its (Gaudi) VLLM instances up, sending large number of requests and then scaling VLLM instances replica count back to 1, while all those instances are still processing the requests.
